### ISSUE TITLE
multi-gpu-dataparrel-cls.py 报错inf

### ISSUE THREAD

**Issue by jiakechong1991** [taishan1994/pytorch-distributed-NLP#4]
**实验条件**：4卡 Ubuntu
**背景**：项目的其他 训练脚本 都能运行复现成功，所以框架环境应该没问题，只有下面的脚本执行有问题
**运行命令**：CUDA_VISIBLE_DEVICES=0,1 python multi-gpu-dataparallel-cls.py 
**报错如下**：
  warnings.warn('Was asked to gather along dimension 0, but all '
![image](https://github.com/taishan1994/pytorch-distributed-NLP/assets/11289552/908732b0-e73b-446a-9cf7-1a653519c810)

**我的排查过程**：
我把模型的输出（logits, label = self.on_step(batch_data)
                loss = self.criterion(logits, label)这两行的结果变量）打印了一下，然后手动计算loss,发现确实是inf
然后我把同样的输入在 model(不打开 数据并行)单卡测试，发现loss正常，
然后把同样的输入在 model(打开 数据并行)上双卡测试,发现loss确实 inf.
一直没有排查出原因，还请大佬指教

**Comment by taishan1994** [comment]
> **实验条件**：4卡 Ubuntu **背景**：项目的其他 训练脚本 都能运行复现成功，所以框架环境应该没问题，只有下面的脚本执行有问题 **运行命令**：CUDA_VISIBLE_DEVICES=0,1 python multi-gpu-dataparallel-cls.py **报错如下**： warnings.warn('Was asked to gather along dimension 0, but all ' ![image](https://user-images.githubusercontent.com/11289552/257757207-908732b0-e73b-446a-9cf7-1a653519c810.png)
> 
> **我的排查过程**： 我把模型的输出（logits, label = self.on_step(batch_data) loss = self.criterion(logits, label)这两行的结果变量）打印了一下，然后手动计算loss,发现确实是inf 然后我把同样的输入在 model(不打开 数据并行)单卡测试，发现loss正常， 然后把通常的输入在 model(打开 数据并行)上双卡测试,发现loss确实 inf. 一直没有排查出原因，还请大佬指教

可能的解决方法：
1、调小学习率。
2、代码中增加梯度裁剪策略。

**Comment by jiakechong1991** [comment]
**实验如下**（单一变量操作）：
1. 把学习率从     learning_rate = 3e-5   改成 1e-5和3e-7  报错不变
2. 把梯度加上裁剪，代码如下：
![image](https://github.com/taishan1994/pytorch-distributed-NLP/assets/11289552/3a0fc465-19f5-4f88-b7c1-4e08f8a5bd89)
报错如下：
![image](https://github.com/taishan1994/pytorch-distributed-NLP/assets/11289552/0c369f0e-de7f-44f0-9c5c-c847cd2a3d2b)

大佬，疑惑一个问题：
1. 同样的模型，同样的输入，为啥单卡时，输出的logits和多卡的输出logits不同呢？dataparallel不就是 每个GPU加载同样的模型吗？

感谢大佬百忙之中指教

**Comment by taishan1994** [comment]
我也不知道是什么原因了。

**Comment by jiakechong1991** [comment]
感谢，我再研究下，如果解决就贴在这里
